### PR TITLE
Adding MonoRes to devel

### PR DESCRIPTION
**Pull request by Vilax**

### CODE REVIEW

File: software/em/xmipp/libraries/data/filters.h [hunk 1]
```diff
@@ -118,6 +118,19 @@ void regionGrowing3D(const MultidimArray< double >& V_in,
...
+/** Region growing with equal values for volumes
+ * @ingroup Filter
+ *
+ * Given a position inside a volume this function grows a region with
+ * (filling_value) until it finds a border. If the point
+ * is outside the volume then nothing is done.
+ *
+ * The growing only considers those voxels with value equal to the selected voxel
+ */
+void regionGrowing3DEqualValue(const MultidimArray<double> &V_in,
```
Comment (delarosatrevin) [inline software/em/xmipp/libraries/data/filters.h]: This new function should have a test.

File: software/em/xmipp/libraries/data/filters.cpp [hunk 2]
```diff
@@ -482,6 +482,93 @@ void regionGrowing3D(const MultidimArray<double> &V_in,
...
+   if (A3D_ELEM(V_in,K,I,J)==bgColor && A3D_ELEM(V_out,K,I,J)==1) { \
+     coord.ii=I; \
+     coord.jj=J; \
+     coord.kk=K; \
+     A3D_ELEM (V_out,coord.kk,coord.ii,coord.jj)=filling_value; \
+     iNeighbours.push_front(coord); \
+    } \
+  }\
+ }\
+  //if (voxel == stop_colour) { \
```
Comment (delarosatrevin) [inline software/em/xmipp/libraries/data/filters.cpp]: Remove unused code.

File: pyworkflow/tests/em/protocols/test_protocols_MonoRes.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
+# **************************************************************************
```
Comment (delarosatrevin) [inline pyworkflow/tests/em/protocols/test_protocols_MonoRes.py]: This file name has Capitals...we do not use that convention.

File: pyworkflow/tests/em/protocols/test_protocols_MonoRes.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                                   symmetry = 'd2',
+                                   minRes = 1,
+                                   maxRes = 100,
+                                   significance = 0.95,
+                                   exact = False,
+                                   filterInput = False,
+                                   trimming = False,
+                                    )
+        self.launchProtocol(MonoRes)
+        self.assertTrue(exists(MonoRes._getExtraPath("MGresolution.vol")), "MonoRes (no split and no mask) has failed")
```
Comment (delarosatrevin) [inline pyworkflow/tests/em/protocols/test_protocols_MonoRes.py]: Check code formatting of this file. Code of lines should not be wide than 80 characters.

File: pyworkflow/em/packages/xmipp3/viewer_resolution_monogenic_signal.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        return [Plotter(fig)]
+                
+             
+    def _plotHistogram(self, param=None):
+        md = MetaData()
+#         path_ = self.protocol._getPath('extra/hist.xmd')
+        md.read(self.protocol._getPath('extra/hist.xmd'))
+        x_axis = []
+        y_axis = []
+#         y_axis_max = 0
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/viewer_resolution_monogenic_signal.py]: Check and remove all unused code in this file.

File: pyworkflow/em/packages/xmipp3/viewer_resolution_monogenic_signal.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+# *  e-mail address 'jmdelarosa@cnb.csic.es'
+# *
+# **************************************************************************
+
+from pyworkflow.gui.plotter import Plotter
+from pyworkflow.protocol.params import LabelParam, FloatParam
+from pyworkflow.viewer import ProtocolViewer, DESKTOP_TKINTER
+from pyworkflow.em.viewer import ChimeraView, ObjectView, DataView
+from protocol_resolution_monogenic_signal import XmippProtMonoRes
+# from pyworkflow.utils import runJob, runCommand
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/viewer_resolution_monogenic_signal.py]: More unused imports.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+    def _methods(self):
+        messages = []
+        if (hasattr(self,'outputParticles')):
+            messages.append('An angular assignment of untilted and tilted particles is carried out [Publication: Not yet]')
+        return messages
+    
+    def _citations(self):
+        return ['Not yet']
+    
+#     def getSummary(self):
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Remove unused code.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+    #--------------------------- INFO functions -------------------------------------------- 
+    def _validate(self):
+        
+        validateMsgs = []
+        if not self.inputVolume.get().hasValue():
+            validateMsgs.append('Please provide input volume.')  
+        return validateMsgs
+
+
+#     def _summary(self):
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Remove unused code.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+            params +=  ' --trimmed %f' % 0
+            
+        self.runJob('xmipp_resolution_monogenic_signal', params)
+        
+        
+    def symmetrizeStep(self, fnVol2Sym):
+        
+        params =  ' -i %s' % fnVol2Sym
+        params +=  ' --sym %s' % self.symmetry.get()
+#        params +=  ' -o %s' % self._getExtraPath('MGresolution2.vol')
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Remove unused code.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        ih.convert(volLocation1, self._getExtraPath('input_volume.vol'))
+        
+        if volLocation2 is not None:
+            ih.convert(volLocation2, self._getExtraPath('input_volume2.vol'))
+            
+        if fnmask is not None:
+            ih.convert(fnmask, self._getExtraPath('input_mask.vol'))
+
+
+#         path_vol = self._getExtraPath() + '/input_volume.vol'
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Remove unused code.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+        volLocation1 = self.inputVolume.get().getFileName()
+        volLocation2 = self.inputVolume2.get().getFileName()
+        fnmask = self.Mask.get().getFileName()
+        
+        convertId = self._insertFunctionStep('convertInputStep', volLocation1, fnmask, volLocation2)
+        deps = []
+                 
+
+        
+#         if self.symmetry.get() not in ['c1', 'C1']:
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Remove unused code.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+
+class XmippProtMonoRes(ProtRefine3D):
+    """    
+    Given a map the protocol assings local resolutions to each pixel of the map.
+    """
+    _label = 'MonoRes - Monogenic Resolution'
+    
+    def __init__(self, *args, **kwargs):
+        ProtRefine3D.__init__(self, *args, **kwargs)
+        #self.stepsExecutionMode = STEPS_PARALLEL
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Remove unused code.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# * GNU General Public License for more details.
+# *
+# * You should have received a copy of the GNU General Public License
+# * along with this program; if not, write to the Free Software
+# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
+# * 02111-1307  USA
+# *
+# *  All comments concerning this program package may be sent to the
+# *  e-mail address 'jmdelarosa@cnb.csic.es'
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Use: scipion@cnb.csic.es

**Comment by pconesa** [comment]
I would like to review this with you @delarosatrevin  if possible.

**Comment by delarosatrevin** [comment]
I'm sorry, @pconesa , I already requested some changes to @Vilax. We can also discuss them tomorrow.

File: pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+
+
+class XmippProtMonoRes(ProtRefine3D):
+    """    
+    Given a map the protocol assings local resolutions to each pixel of the map.
+    """
+    _label = 'MonoRes - Monogenic Resolution'
+    
+    def __init__(self, *args, **kwargs):
```
Comment (delarosatrevin) [inline pyworkflow/em/packages/xmipp3/protocol_resolution_monogenic_signal.py]: Default constructor is not needed.

**Comment by pconesa** [comment]
So far so good, @Vilax. Resolution and color scale looks consistent with histogram. But I'm getting a weird error in the viewer, for the 2 halves protocol, requesting the coloured slices.

Traceback (most recent call last):
  File "/home/pablo/desarrollo/scipion/software/lib/python2.7/lib-tk/Tkinter.py", line 1486, in __call__
    return self.func(*args)
  File "/home/pablo/desarrollo/scipion/pyworkflow/gui/form.py", line 948, in _visualizeVar
    self.visualizeCallback(self.paramName)
  File "/home/pablo/desarrollo/scipion/pyworkflow/viewer.py", line 306, in _visualizeParam
    v.show()
  File "/home/pablo/desarrollo/scipion/pyworkflow/gui/plotter.py", line 188, in show
    self.tightLayout()
  File "/home/pablo/desarrollo/scipion/pyworkflow/gui/plotter.py", line 185, in tightLayout
    plt.tight_layout()
  File "/home/pablo/desarrollo/scipion/software/lib/python2.7/site-packages/matplotlib-1.3.1-py2.7-linux-x86_64.egg/matplotlib/pyplot.py", line 1255, in tight_layout
    fig.tight_layout(pad=pad, h_pad=h_pad, w_pad=w_pad, rect=rect)
  File "/home/pablo/desarrollo/scipion/software/lib/python2.7/site-packages/matplotlib-1.3.1-py2.7-linux-x86_64.egg/matplotlib/figure.py", line 1605, in tight_layout
    rect=rect)
  File "/home/pablo/desarrollo/scipion/software/lib/python2.7/site-packages/matplotlib-1.3.1-py2.7-linux-x86_64.egg/matplotlib/tight_layout.py", line 325, in get_tight_layout_figure
    max_nrows = max(nrows_list)
ValueError: max() arg is an empty sequence

The second time, it does nos fail and shows 3 figures, but one is empty.

**Pull request merged**